src/librustc_codegen_llvm/attributes.rs

```diff
@@ -310,6 +310,16 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
         .collect::<Vec<String>>()
         .join(",");
 
+    match codegen_fn_attrs.instruction_set.as_ref() {
+        Some(InstructionSetAttr::ArmA32) => {
+            features.push_str("-thumb-mode");
+        },
+        Some(InstructionSetAttr::ArmT32) => {
+            features.push_str("+thumb-mode");
```

Comment:
This won't work if there are any `#[target_features]` (generates an invalid string). I suggest `.chain`ing the thumb mode feature onto the iterator above.